poseidon/src/permutation.rs

```diff
@@ -93,34 +104,40 @@ pub fn half_rounds<F: Field, SC: SpongeConstants>(
         {
             state[i].add_assign(x);
         }
+
         for state_i in state.iter_mut() {
             *state_i = sbox::<F, SC>(*state_i);
         }
-        let res = apply_mds_matrix::<F, SC>(params, state);
-        res.iter().enumerate().for_each(|(i, x)| {
-            state[i] = *x;
-        });
+
+        apply_mds_matrix::<F, SC>(params.mds, state);
     }
 }
 
-pub fn poseidon_block_cipher<F: Field, SC: SpongeConstants>(
-    params: &ArithmeticSpongeParams<F>,
-    state: &mut Vec<F>,
+pub fn poseidon_block_cipher<F: Field, SC: SpongeConstants, const FULL_ROUNDS: usize>(
+    params: &ArithmeticSpongeParams<F, FULL_ROUNDS>,
+    state: &mut [F],
 ) {
     if SC::PERM_HALF_ROUNDS_FULL == 0 {
         if SC::PERM_INITIAL_ARK {
-            for (i, x) in params.round_constants[0].iter().enumerate() {
-                state[i].add_assign(x);
-            }
+            // maintaining previous invariants
+            assert!(params.round_constants[0].len() <= state.len());
```

Comment:
Previously, if `round_constants[0]` was larger than `state`, `state[i]` would panic with an out-of-bounds access